consensus/hotstuff/README.md

```diff
@@ -78,58 +78,62 @@ The figure below illustrates the dependencies of the core components and informa
+with views strictly smaller than the last finalized block, the chain decomposes into multiple disconnected trees, in graph theory, such structure is a forest, to separate general graph-theoretical 
+concepts from the concrete blockchain application, `LevelledForest` refers to blocks as graph `vertices` and to a block's view number as `level`, 
+the `LevelledForest` is an in-memory data structure to store and maintain a levelled forest, it provides functions to add vertices, query vertices by their ID (block's hash), query vertices by level, query the children of a vertex, and prune vertices by level (remove them from memory).
+`Forks` tracks the finalized blocks and triggers finalization events whenever it observes a valid extension to the chain of finalized blocks.
 * `Validator` validates the HotStuff-relevant aspects of
    - QC: total weight of all signers is more than 2/3 of committee weight, validity of signatures, view number is strictly monotonously increasing
-   - block proposal: from designated primary for the block's respective view, contains proposer's vote for its own block, QC in block is valid
-   - vote: validity of signature, voter is has positive weight 
+   - TC: total weight of all signers is more than 2/3 of committee weight, validity of signatures, proof for entering view. 
+   - block proposal: from designated primary for the block's respective view, contains proposer's vote for its own block, QC in block is valid, TC if needed is valid.
+   - vote: validity of signature, voter is has positive weight
 * `VoteAggregator` caches votes on a per-block basis and builds QC if enough votes have been accumulated.
-* `Voter` tracks the view of the latest vote and determines whether or not to vote for a block (by calling `forks.IsSafeBlock`)
-* `Committee` maintains the list of all authorized network members and their respective weight on a per-block basis. Furthermore, the committee contains the primary selection algorithm. 
+* `TimeoutAggregator` caches timeouts on a per-view basis and builds TC if enough timeouts have been accumulated. Performs validation and verification of timeouts.
+* `Replicas` maintains the list of all authorized network members and their respective weight on a per-view basis. Maintains a static list which changes only between epochs. Furthermore, the `Replicas` contains the primary selection algorithm.
+* `DynamicCommittee` maintains the list of all authorized network members and their respective weight on a per-block basis. Extends `Replicas`.
 * `BlockProducer` constructs the payload of a block, after the HotStuff core logic has decided which fork to extend 
 
 # Implementation
 
-We have translated the Chained HotStuff protocol into a state machine shown below. The state machine is implemented 
-in `EventHandler`.
+We have translated the HotStuff protocol into a state machine shown below. The state machine is implemented in `EventHandler`.
 
 ![](/docs/StateMachine.png)
```

Comment:
![IMG_93209EDD4263-1](https://user-images.githubusercontent.com/10557821/215582526-f907d75b-f3dd-4a25-b7d6-1b0adaca202c.jpeg)

Suggesting a minor change to this diagram - we check whether the block is for current view before voting or proposing.

Reply:
I think this check is in the right place. That check circled in red is about next line: https://github.com/onflow/flow-go/blob/feature/active-pacemaker/consensus/hotstuff/eventhandler/event_handler.go#L181

Reply:
the first check is part of "produce vote and send to message hub" step and it's incorporated into that, not as separate statement but other one is a separate statement in this diagram, since it's a logical statement which means that proposing makes sense only if the block processing has changed the active view, otherwise not.

Reply:
We want to provide the figure, to help understand the code. So if this figure is not reasonably clear (which Jordan's comment indicates it is not), then we should be improving the figure. 

I feel the "produce vote" and "try to propose" leave out some details that are relatively important for the algorithm to make sense. 

Suggestion:
![StateMachine](https://user-images.githubusercontent.com/26322303/217103806-424334d4-fd9f-4772-bec7-ddf06a14378d.png)